src/lib.rs

```diff
@@ -17,46 +17,171 @@
+const INITIAL_FUNDS_ALICE: u64 = 1000;
+/// How often we mine a block.
+const BLOCK_TIME_SECS: u64 = 1;
+/// Poll interval when checking if the wallet has synced with monerod.
+const WAIT_WALLET_SYNC_MILLIS: u64 = 1000;
+
+/// RPC client for `monerod` and `monero-wallet-rpc`.
 #[derive(Debug)]
-pub struct Client {
+pub struct Client<'c> {
+    container: Container<'c, clients::Cli, GenericImage>,
     pub wallet: wallet::Client,
     pub monerod: monerod::Client,
 }
 
-impl Default for Client {
-    fn default() -> Self {
+impl<'c> Client<'c> {
+    pub fn new(
+        docker_client: &'c clients::Cli,
+        monerod_rpc_port: u16,
+        wallet_rpc_port: u16,
+    ) -> Self {
+        let image = images::generic::GenericImage::new("xmrto/monero")
+            .with_mapped_port(Port {
+                local: monerod_rpc_port,
+                internal: 28081,
+            })
+            .with_mapped_port(Port {
+                local: wallet_rpc_port,
+                internal: 28083,
+            })
+            .with_entrypoint("")
+            .with_args(vec![
+                "/bin/bash".to_string(),
+                "-c".to_string(),
+                "monerod --confirm-external-bind --non-interactive --regtest --rpc-bind-ip 0.0.0.0 --rpc-bind-port 28081 --no-igd --hide-my-port --fixed-difficulty 1 --rpc-payment-allow-free-loopback --data-dir /monero & \
+                monero-wallet-rpc --log-level 4 --daemon-address localhost:28081 --confirm-external-bind --disable-rpc-login --rpc-bind-ip 0.0.0.0 --rpc-bind-port 28083  --wallet-dir /monero/".to_string(),
+            ])
+            .with_wait_for(WaitFor::LogMessage { message: "You are now synchronized with the network. You may now start monero-wallet-cli".to_string(), stream: Stream::StdOut });
+        let container = docker_client.run(image);
         Self {
-            wallet: wallet::Client::localhost(MONERO_WALLET_RPC_PORT)
+            container,
+            wallet: wallet::Client::localhost(wallet_rpc_port)
                 .expect("failed to create wallet client"),
-            monerod: monerod::Client::localhost(MONEROD_RPC_PORT)
+            monerod: monerod::Client::localhost(monerod_rpc_port)
                 .expect("failed to create monerod client"),
         }
     }
+
+    /// Constructor that generates random port numbers for the local port
+    /// mapping of `monerod` and `monero-wallet-rpc`.
+    pub fn new_with_random_local_ports(cli: &'c clients::Cli) -> Self {
+        let mut rng = rand::thread_rng();
+        let monerod_port: u16 = rng.gen_range(1024, u16::MAX);
+        let wallet_port: u16 = rng.gen_range(1024, u16::MAX);
+
+        Client::new(cli, monerod_port, wallet_port)
+    }
+
+    /// Initialise by creating a wallet, generating some `blocks`, and starting
+    /// a miner thread that mines to the primary account.
+    pub async fn init(&self, blocks: u32) -> Result<()> {
```

Comment:
Having multiple `init` methods that all just take `&self` doesn't make use of the type system really well.

I'd assume you would only want one of those to actually be called. Could make multiple constructors instead so one has to choose which one to run.

Reply:
Good idea, thanks.

Reply:
Ok, I messed around with this for ages - way too long. In the end I removed the second constructor, removed one of the inits and commented the other saying its probably not the one you want (kept it because it is useful for testing but `#[cfg(test)] doesn't help since we test it in `/tests/`. @bonomat I removed the default init that I argue for having in there :)